File: lab4/cpp/source/philosopher.cpp
#include "philosopher.hpp"

#include "dining_philosophers_list.hpp"

#include <chrono>
#include <iostream>

#define RSEED 42

#define MAX_RAND_TIME 1000
#define EATING_TIME_OFFSET 200
#define THINKING_TIME_OFFSET 500

namespace jpp {

philosopher::philosopher(const uint16_t id, const uint16_t num_philosophers)
: _id(id), _left_cutlery_id(id), _right_cutlery_id((id + 1) % num_philosophers) {}

uint16_t philosopher::id() const {
    return this->_id;
}

std::string philosopher::to_string() const {
    return "(C" + std::to_string(this->_left_cutlery_id)
         + "-P" + std::to_string(this->_id)
         + "-C" + std::to_string(this->_right_cutlery_id)
         + ":M" + std::to_string(this->_meals_eaten) + ")";
    // TODO: use std::format
}

void philosopher::_start(
    const uint16_t max_meals,
    std::vector<std::unique_ptr<std::binary_semaphore>>& cutlery_semaphore_list,
    dining_philosophers_list& dinig_list
) {
    static std::mt19937 gen32;
    static std::uniform_int_distribution<uint16_t> uniform{1, MAX_RAND_TIME - 1};

    while (true) {
        // think
        std::this_thread::sleep_for(
            std::chrono::milliseconds(uniform(gen32) + THINKING_TIME_OFFSET)
        );

        // acquire cutlery
        cutlery_semaphore_list[this->_left_cutlery_id]->acquire();
        cutlery_semaphore_list[this->_right_cutlery_id]->acquire();

        // add to dining list
        this->_meals_eaten++;
        dinig_list.add_philosopher(*this);

        // eat
        std::this_thread::sleep_for(
            std::chrono::milliseconds(uniform(gen32) + EATING_TIME_OFFSET)
        );

        // remove from dining list
        dinig_list.remove_philosopher(this->_id);

        // release cutlery
        cutlery_semaphore_list[this->_left_cutlery_id]->release();
        cutlery_semaphore_list[this->_right_cutlery_id]->release();


        if (bool finished_eating = this->_meals_eaten == max_meals; finished_eating) {
            std::cout << "[Finished eating: id = " << this->_id << "]\n";
            return;
        }
    }
}

void philosopher::start_thread(
    const uint16_t p_id,
    const uint16_t num_philosophers,
    const uint16_t max_meals,
    std::vector<std::unique_ptr<std::binary_semaphore>>& cutlery_semaphore_list,
    dining_philosophers_list& dinig_list
) {
    philosopher p{p_id, num_philosophers};
    p._start(max_meals, cutlery_semaphore_list, dinig_list);
}

} // namespace jpp

File: lab4/cs/DiningPhilosophers/Philosopher.cs
namespace DiningPhilosophers;

using System;
using System.Threading;

public class Philosopher {
    public Philosopher(byte id, byte numPhilosophers = 1) {
        this._id = id;
        this._leftCutleryId = id;
        this._rightCutleryId = (byte)((id + 1) % numPhilosophers);
    }

    public ulong Id {
        get { return this._id; }
    }

    public override string ToString()
        => $"(C{this._leftCutleryId}-P{this._id}-C{this._rightCutleryId}:M{this._mealsEaten})";

    public static void StartThread(
        byte pId,
        byte numPhilosophers,
        byte maxMeals,
        ref List<Semaphore> cutlerySemaphoreList,
        ref DiningPhilosophersList diningList
    ) {
        Philosopher p = new Philosopher(pId, numPhilosophers);

        while (true) {
            // think
            Thread.Sleep(Philosopher.s_random.Next(Philosopher.MaxRandTime) + Philosopher.ThinkingTimeOffset);

            // acquire cutlery
            cutlerySemaphoreList[p._leftCutleryId].WaitOne();
            cutlerySemaphoreList[p._rightCutleryId].WaitOne();

            // add to dining list
            p._mealsEaten++;
            diningList.AddPhilosopher(p);

            // eat
            Thread.Sleep(Philosopher.s_random.Next(Philosopher.MaxRandTime) + Philosopher.EatingTimeOffset);

            // remove from dining list
            diningList.RemovePhilosopher(p._id);

            // release cutlery
            cutlerySemaphoreList[p._leftCutleryId].Release();
            cutlerySemaphoreList[p._rightCutleryId].Release();

            if (p._mealsEaten == maxMeals) {
                Console.WriteLine($"[Finished eating: id = {p._id}]");
                return;
            }
        }
    }

    private readonly byte _id;
    private readonly byte _leftCutleryId;
    private readonly byte _rightCutleryId;
    private byte _mealsEaten = 0;

    private static Random s_random = new Random();
    private static readonly int MaxRandTime = 1000;
    private static readonly int EatingTimeOffset = 200;
    private static readonly int ThinkingTimeOffset = 500;
}
